rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,241 @@
+# Distributed DNS Load Balancing
+
+- Feature Name: Distributed DNS Load Balancing
+- Start Date: 2024-01-17
+- RFC PR: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/pull/55)
+- Issue tracking: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/issues/56)
+
+# Summary
+[summary]: #summary
+
+Enable the DNS Operator to manage DNS for one or more hostnames shared across one or more clusters.  Remove the requirement for a central "hub" cluster or multi-cluster control plane. Increase the robustness of the DNS Health checks by executing them from all the available clusters.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the DNS operator has limited functionality for a single cluster, and none for multiple clusters without OCM. Having the DNS operator enabled to manage DNS in these situations significantly eases onboarding users to leveraging DNSPolicy across both single and multi-cluster deployments.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- ![Use Cases](0008-distributed-dns-assets%2Fdynamic-dns-use-cases.jpg)
+- ![P2P Health Checks](0008-distributed-dns-assets%2Fdynamic-dns-health-checks.jpg)
+- ![DNS Pruning](0008-distributed-dns-assets%2Fdynamic-dns-dns-pruning.jpg)
+- ![DNS Controller](0008-distributed-dns-assets%2Fdynamic-dns-dns-controller.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster and Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret. If a provider secret reference the same zone(s) in the same backing DNS Provider for the same host, it will be treated as a distributed DNS target. In each cluster, when a matching listener host defined in the gateway is found that matches one of the available zones in the provider the addresses from that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways addresses to the same zone file, without destroying each other's values.
+
+### Health Checks
+If health checks are defined in the DNS Policy, these will be created on all clusters. When a health check is failing, a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
```

Comment:
```suggestion
If health checks are defined in the DNS Policy, these will be created and executed from the same cluster where the DNSPolicy is defined and target each cluster using that listener host name. When a health check is failing, a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the failure threshold (66%) by default but overridable via flag on the controller  , that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP. While 66% is a little arbitrary, it can be changed and gives a decent base line of 2 out of every 3 clusters failing before an IP is removed.
```

Reply:
If it is configured in the controller, it needs to be configured on all controllers with the same value right, otherwise it might cause issues

Reply:
yes although that would be in a purposeful override scenerio. Which we should def document